### PR TITLE
feat(common): Write missing cases

### LINKED ISSUE #9: SES integration
Add remaining cases: ConfirmSignUp, ForgotPassword, ConfirmForgotPassword

### PR DESCRIPTION
**Pull request by gaebel**
I wrote the missing cases for the SES integration (Confirm sign up, forgot password, confirm forgot password) and rewrote the functions to use onResult for better readability. [ fixes #9 ]

@benjohnde Any idea how we can test this?

### CODE REVIEW

**Comment by gaebel** [comment]
@benjohnde Probably too much effort to make this testable. Should we skip this? 🙈

File: auth/src/commonMain/kotlin/com/liftric/Auth.kt [hunk 1]
```diff
@@ -48,14 +49,37 @@ interface Auth {
...
+    suspend fun forgotPassword(username: String): Result<CodeDeliveryDetails>
+
+    /**
+     * Confirms sign up of a new user
+     * @param username The username
+     * @param username The new password that was sent to the users' delivery medium
+     * @param confirmationCode The confirmation code that was sent to the users' delivery medium
+     * @return Result object containing Unit on success or an error on failure
+     */
+    suspend fun confirmforgotPassword(username: String, password: String, confirmationCode: String): Result<Unit>
```
Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/Auth.kt:70]: The `f` in `confirmforgot` should be uppercase.

File: auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt [hunk 2]
```diff
@@ -38,7 +35,7 @@ open class AuthHandler(private val configuration: Configuration) : Auth {
         password: String,
         attributes: List<UserAttribute>?
     ): Result<SignUpResponse> {
-        val response = request(
+        return request(
```
Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:38]: You may omit the return 🤔

File: auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt [hunk 3]
```diff
@@ -49,23 +46,39 @@ open class AuthHandler(private val configuration: Configuration) : Auth {
                     UserAttributes = attributes ?: listOf()
                 )
             )
-        )
-        return if (response.isSuccess) {
+        ).onResult {
             try {
-                Result.success(parse(SignUpResponse.serializer(), response.getOrNull()!!))
+                Result.success(parse(SignUpResponse.serializer(), it))
```
Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:51]: That's just so much better with the `onResult` invocation 🎉 .

File: auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt [hunk 4]
```diff
@@ -49,23 +46,39 @@ open class AuthHandler(private val configuration: Configuration) : Auth {
...
-        } else {
-            Result.failure(response.exceptionOrNull()!!)
+        }
+    }
+
+    override suspend fun confirmSignUp(
+        username: String,
+        confirmationCode: String
+    ): Result<Unit> {
+        return request(
```
Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:62]: Not quite sure, potentially omit the `return` 🤔 .

File: auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt [hunk 5]
```diff
@@ -49,23 +46,39 @@ open class AuthHandler(private val configuration: Configuration) : Auth {
...
+            Result.success(Unit)
         }
     }
 
     override suspend fun signIn(
         username: String,
         password: String
     ): Result<SignInResponse> {
-        val response = request(
+        return request(
```
Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:81]: see above

File: auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt [hunk 6]
```diff
@@ -159,26 +148,74 @@ open class AuthHandler(private val configuration: Configuration) : Auth {
...
+        ).onResult {
+            try {
+                Result.success(parse(CodeDeliveryDetails.serializer(), it))
+            } catch (e: SerializationException) {
+                Result.failure(e)
+            }
+        }
+    }
+
+    override suspend fun confirmforgotPassword(
```
Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:177]: see above (the f in `forgot` should be uppercased)

File: auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt [hunk 7]
```diff
@@ -159,26 +148,74 @@ open class AuthHandler(private val configuration: Configuration) : Auth {
...
+        username: String,
+        password: String,
+        confirmationCode: String
+    ): Result<Unit> {
+        return request(
```
Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:182]: see above

File: auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt [hunk 8]
```diff
@@ -159,26 +148,74 @@ open class AuthHandler(private val configuration: Configuration) : Auth {
...
+                    confirmationCode
+                )
+            )
+        ).onResult {
+            Result.success(Unit)
+        }
+    }
+
+    override suspend fun signOut(accessToken: String): Result<Unit> {
+        return request(
```
Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:199]: see above

File: auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt [hunk 9]
```diff
@@ -206,7 +243,7 @@ open class AuthHandler(private val configuration: Configuration) : Auth {
                 )
                 body = payload
             }
-            return if (response.status.value == 200) {
+            return if (response.status == HttpStatusCode.OK) {
```
Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:246]: 👍

Comment (benjohnde) [inline auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:246]: see above (omitting return)

**Comment by benjohnde** [comment]
> @benjohnde Probably too much effort to make the functions with email confirmation testable. Should we omit the tests? 🙈

I would omit the tests as for these cases we would need a fully configured SES with some Cognito User Pool. As those functions are implemented according to the Cognito API spec: full ack to skip those nifty heavy integration tests.

File: README.md [hunk 10]
```diff
@@ -131,16 +131,16 @@ Invokes password forgot and sends a confirmation code the the users' delivery me
...
 ```
 
 #### Confirm Forgot Password
 
 Confirms forgot password.
 
 ```kotlin
-changePassword(username = "USERNAME", password = "NEW_PASSWORD_FROM_DELIVERY_MEDIUM",
-               confirmationCode = "CODE_FROM_DELIVERY_MEDIUM"): Result<Unit>
+onfirmForgotPassword(username = "USERNAME", password = "NEW_PASSWORD_FROM_DELIVERY_MEDIUM",
```
Comment (benjohnde) [inline README.md:142]: Oweee missing `c` :)

File: auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt (hunk 9 above)
Comment (gaebel) [reply auth/src/commonMain/kotlin/com/liftric/AuthHandler.kt:246]: A 'return' expression required in a function with a block body ('{...}')

**Comment by benjohnde** [comment]
LGTM, feel free to merge!

**Review by benjohnde — APPROVED** [review]

**Pull request merged**